src/hotspot/share/services/nmtPreInit.hpp

```diff
@@ -0,0 +1,353 @@
+//   around a lot.
+// - However, *surviving* allocations (allocations not freed immediately) are much rarer:
+//   typically only about 300-500. Again, mainly depending on the number of VM args.
+// - There are a few cases of pre-to-post-init reallocs where pre-init allocations get
+//   reallocated after NMT initialization. Those we need to handle with special care (see
+//   NMTPreInit::handle_realloc()). Because of them we need to store allocation size with
+//   every pre-init allocation.
+
+// For the lookup table, design considerations are:
+//   - lookup speed is paramount since lookup is done for every os::free() call.
+//   - insert/delete speed only matters for VM startup - after NMT initialization the lookup
+//     table is readonly
+//   - memory consumption of the lookup table matters since we always pay for it, NMT on or off.
+//   - Obviously, nothing here can use *os::malloc*. Any dynamic allocations - if they cannot
+//     be avoided - should use raw malloc(3).
+//
+// We use a basic open hashmap, dimensioned generously - hash collisions should be very rare.
+//  The table is customized for holding malloced pointers. One main point of this map is that we do
+//  not allocate memory for the nodes themselves. Instead we piggy-back on the user allocation:
+//  the hashmap entry structure precedes, as a header, the malloced block. That way we avoid extra
+//  allocations just to hold the map nodes. This keeps runtime/memory overhead as small as possible.
+
+struct NMTPreInitAllocation {
+  NMTPreInitAllocation* next;
+  const size_t size; // (inner) payload size without header
+  // <-- USER ALLOCATION (PAYLOAD) STARTS HERE -->
+
+  NMTPreInitAllocation(size_t size) : next(NULL), size(size) {};
+
+  // Returns start of the user data area
+  void* payload()             { return this + 1; }
+  const void* payload() const { return this + 1; }
+
+  // These functions do allocate/free a raw malloc block of given payload size,
+  //  preceded with a NMTPreInitAllocation header.
+  static NMTPreInitAllocation* do_alloc(size_t payload_size);
+  static NMTPreInitAllocation* do_reallocate(NMTPreInitAllocation* old, size_t new_payload_size);
+  static void do_free(NMTPreInitAllocation* p);
+};
+
+class NMTPreInitAllocationTable {
+
+  // table_size: keep table size a prime and the hash function simple; this
+  //  seems to give a good distribution for malloce'd pointers on all our libc variants.
+  // 8000ish is really plenty: normal VM runs have ~500 pre-init allocations to hold,
+  //  VMs with insanely long command lines maybe ~700-1000. Which gives us an expected
+  //  load factor of ~.1. Hash collisions should be very rare.
+  // ~8000 entries cost us ~64K for this table (64-bit), which is acceptable.
+  static const int table_size = 7919;
+
+  NMTPreInitAllocation* _entries[table_size];
+
+  typedef int index_t;
+  const index_t invalid_index = -1;
+
+  static unsigned calculate_hash(const void* p) {
+    uintptr_t tmp = p2i(p);
```

Comment:
malloc memory usually is 2-machine word aligned, maybe tmp = tmp >> LP64_ONLY(4) NOT_LP64(3) can result better hash distribution?

Reply:
I thought so too at first, but it is not necessary. The hash function uses all input bits:
```
    uintptr_t tmp = p2i(p);
    unsigned hash = (unsigned)tmp
                     LP64_ONLY( ^ (unsigned)(tmp >> 32));
```
p2i is lossless since input and result have the same size.
Then,
- for 32-bit, it does not matter since unsigned is the same size as uintptr_t and we lose no bits
- for 64-bit, we xor the upper half of the 64-bit input value into the lower half; again, all input bits count toward the result. We don't gain anything by shifting, we would just exchange the lower 2 bits always being zero against the upper 2 bits always being zero.

---

BTW, I experimented with different hash functions, e.g. http://www.concentric.net/~Ttwang/tech/inthash.htm, but did not really get a better distribution. This somewhat mirrors my experiences when I tried to optimize your hash function in the MallocSiteTable. Seems malloc return value is already "random enough". I coupled it with a crooked table size though, made it a prime.